calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,258 @@
+/// a) consist of only writes to 2 different non-combination cells (let's
+/// call them cell1 and cell2) or the group's done port
+/// b) must have cell2.go = cell1.done assignment
+/// c) group[done] = cell2.done
+pub struct GroupToSeq {
+    ///Maps names of group to the sequences that will replace them
+    group_seq_map: BTreeMap<ir::Id, ir::Control>,
+}
+
+impl Named for GroupToSeq {
+    fn name() -> &'static str {
+        "group-to-seq"
+    }
+
+    fn description() -> &'static str {
+        "split groups under correct conditions"
+    }
+}
+
+//If asmt is a write to a cell named name returns Some(name).
+//If asmt is a write to a group port, returns None.
+fn writes_to_cell(asmt: &ir::Assignment) -> Option<ir::Id> {
+    match &asmt.dst.borrow().parent {
+        ir::PortParent::Cell(cell) => {
+            Some(cell.upgrade().borrow().name().clone())
+        }
+        ir::PortParent::Group(_) => None,
+    }
+}
+
+//Given asmt a.go = b.done, return (a1, a2), where a1 is group[done] = b.done,
+//and a2 is a.go = 1'd1.
+fn split_go_done(
+    builder: &mut ir::Builder,
+    asmt: ir::Assignment,
+    group: ir::WRC<ir::Group>,
+) -> (ir::Assignment, ir::Assignment) {
+    let con = builder.add_constant(1, 1);
+    let src = ir::Port {
+        name: ir::Id::new("const 1", None),
+        width: 1,
+        direction: ir::Direction::Output,
+        parent: ir::PortParent::Cell(ir::WRC::from(&con)),
+        attributes: ir::Attributes::default(),
+    };
+    let src_ref = Rc::new(RefCell::new(src));
+
+    let dst = ir::Port {
+        name: ir::Id::new("done", None),
+        width: 1,
+        direction: ir::Direction::Input,
+        parent: ir::PortParent::Group(group),
+        attributes: ir::Attributes::default(),
+    };
+    let dst_ref = Rc::new(RefCell::new(dst));
+    (
+        builder.build_assignment(dst_ref, asmt.src, ir::Guard::True),
+        builder.build_assignment(asmt.dst, src_ref, ir::Guard::True),
+    )
+}
+
+//Given asmt old_group[done] = guard? a.done, returns group[done] = guard? a.done.
+fn rename_group_done(
+    builder: &mut ir::Builder,
+    asmt: ir::Assignment,
+    group: ir::WRC<ir::Group>,
+) -> ir::Assignment {
+    let dst = ir::Port {
+        name: ir::Id::new("done", None),
+        width: 1,
+        direction: ir::Direction::Input,
+        parent: ir::PortParent::Group(group),
+        attributes: ir::Attributes::default(),
+    };
+    let dst_ref = Rc::new(RefCell::new(dst));
+    builder.build_assignment(dst_ref, asmt.src, *asmt.guard)
+}
+
+//Given asmt a.go = !a.done ? 1'd1 returns a.go = 1'd1.
+fn make_go_const(
+    builder: &mut ir::Builder,
+    asmt: ir::Assignment,
+) -> ir::Assignment {
+    let con = builder.add_constant(1, 1);
+    let src = ir::Port {
+        name: ir::Id::new("const 1", None),
+        width: 1,
+        direction: ir::Direction::Output,
+        parent: ir::PortParent::Cell(ir::WRC::from(&con)),
+        attributes: ir::Attributes::default(),
+    };
```

Comment:
Basically any use of `ir::Port` is probably wrong. You should only use the `Cell::get` and `Group::get` to get `ir::Port` objects